### PR TITLE
Refactor hybrid learning card grid

### LINKED ISSUE #8983: Write unit tests for HybridLearningCardGrid

### LINKED ISSUE #9004: Refactor HybridLearningCardGrid to make less brittle and improve readability
The HybridLearningCardGrid component handles a variety of different card displays and manages mobile responsiveness. The code is also inconsistent in places, especially around the conditions that determine the card style that is rendered in the grid. 

The following problems should be solved by the refactor: 
- [ ] update all places that check which page is the current page to consistently use `PageNames`
- [ ] brittleness in conditions/checking
          - can the conditional checks themselves be improved for stability? If so, refactor these
          - otherwise, they should at least be easier to read (potentially turn into computed props rather than long `&&` `||` etc
- [ ] comments should be added for readability and understanding which cards are rendered by which sections of code
- [ ] any other cleanup welcome

Important also when refactoring the conditions to ensure that the mobile view doesn't display two very long (vertically) very squished (horizontally) cards, a regression that popped up earlier, although I cannot find the issue now...

### PR DESCRIPTION
**Pull request by marcellamaki**
## Summary

Removes `HybridLearningCardGrid`
Fixes #9004
Fixes #8983 

`HybridLearningCardGrid` was clearly doing too much, and not doing any of it especially well. I approached refactoring by "unabstracting" to hopefully make things simpler. 

**For the Library and Topics pages**, which continued to have very similar layouts and variation only in the number of columns, I created a new component called `LibraryAndChannelBrowserMainContent` with just the elements from `HybridLearningCardGrid` that were needed for these two pages.  This new component uses the existing `CardGrid` component for grid displays as needed. It also has associated tests 🎉  
(Separately, I've gotten extremely reasonable/helpful feedback from @sairina that the mismatch between channel cards and `TopicsPage` is confusing. This is a first step in moving towards more consistent naming.) 

**For the Bookmarks page and lessons playlist page**, there was no need for grids at all, as the content is displayed in "list" style cards. The relevant cards were added directly to their pages. 

This approach I think remedies my previous poor decision of `HybridLearningCardGrid` having to keep track of which page it was on, to know what to display, and overall makes the card displays less brittle. It also reduces the number of levels which props and events need to traverse on some pages, which is nice! I do not think this makes the code any _shorter_, but hopefully this is a step in the right direction towards readability and maintainability. 

**Suggestions for other improvements, and whether they are blocking or follow-up, would be welcome and helpful!**

…

## Reviewer guidance
Because this refactor touched a lot of the learn pages, I have done my best to layout what would need to be reviewed for manual QA. 

**Lesson page**
(Home Page > Your Classes > Your lessons > [Lesson]
- [ ] list of lesson resources should display 
- [ ] clicking on the lesson should open the resource 
- [ ] using the "back" arrow within the content renderer should return the user to the lesson page 

**Library Page**
- [ ] list of resent resources should display when there are recent resources 
- [ ] list of search results should display when there are search results
- [ ] users should be able to toggle between list and grid view 
- [ ] users should be able to open the info panel with the "i" button on each card 
- [ ] users should be able to open the copies modal on resources that have multiple copies
- [ ] clicking on the content should open the resource 
- [ ] using the "back" arrow within the content renderer should return the user to the library page

**Topics/Channel Browsing Page**
- [ ] list of folders should display when there are folders
- [ ] list of content items within a folder should display when at that level within a folder
- [ ] list of search results should display when there are search results
- [ ] users should be able to open the info panel with the "i" button on each card 
- [ ] clicking on the content should open the resource 
- [ ] using the "back" arrow within the content renderer should return the user to the topics page

**Bookmarks page**
- [ ] list of bookmarks should display when there are bookmarks
- [ ] users should be able to open the info panel with the "i" button on each card 
- [ ] users should be able to remove the bookmark with the "x" button on each card 
- [ ] removing the bookmark should remove the bookmark from it's "current place in line" 
- [ ] clicking on the content should open the resource 
- [ ] using the "back" arrow within the content renderer should return the user to the bookmarks page

**Overall**
- [ ] for library and topic/channel pages, the grid should update the number of columns on resize to medium and small screen widths
- [ ] for all mobile devices, a single column of cards should display (never that weird bug with two columns of very narrow skinny cards that was popping up)

**tl;dr --manual QA should elicit a response somewhat like "Why has Marcella spent several days on this code? nothing is different??"**

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by radinamatic** [comment]
@pcenov Let's give this a walk-through with built assets (EXE, DEB & APK), thank you!

**Comment by pcenov** [comment]
Tested with the Buildkite assets in Windows 10 (Chrome, Firefox, Edge, Internet Explorer), Ubuntu (Chrome, Firefox) and Android (the app and Chrome) as well as in Safari browser on iPad Air simulator. Mostly everything besides the Search and the 'i' icon in IE is fully functional, I was able to identify only the following issues:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Library Page - Search by keyword or filter returns no results if the user is not signed in or hasn't viewed any resources yet | |
:-------------------------:|:-------------------------:
| Library Page - Search returns incorrect number of results and View more is not functioning (an extended version of the first issue, the Recent items are displayed as search results) - Steps to replicate: 1. Create a new user. 2. Go to Library and open a resource so that there's something in the Recent section 3. Filter by Language: English | |
:-------------------------:|:-------------------------:
| Library Page - Search with applied filter returns unfiltered results|  |
:-------------------------:|:-------------------------:
| Internet Explorer Browser only - User is not able to open the info panel with the "i" button on each card |  |

**Comment by marcellamaki** [comment]
Thanks so much @pcenov -- will address these today!

**Comment by marcellamaki** [comment]
@pcenov -- the first three errors seemed to all be due to me accidentally passing a wrong piece of data, so I think I've solved that and will update shortly 🎉  

For
> Internet Explorer Browser only - User is not able to open the info panel with the "i" button on each card

one question: is this issue present on all pages, or is it just the bookmarks page?  i.e. is it broken on the Library page too? 

Screen Shot 2022-02-01 at 10 44 52 AM

Thank you!

**Comment by pcenov** [comment]
> one question: is this issue present on all pages, or is it just the bookmarks page? i.e. is it broken on the Library page too?

Yup, it's persistent on all pages.

**Comment by marcellamaki** [comment]
Hi @pcenov and @radinamatic -- I've pushed the changes to fix the first 3 issues. I tested the fourth issue in a virtual IE browser and was able to replicate it. I checked out the current version of 0.15.x and it seems to exist there too. So, @pcenov I think you might have found an existing bug. At first I thought perhaps it was from the card refactor Devon and I worked on that was just recently merged, but it seems to be existing on both the bookmarks cards (which were refactored) and the library page (which was not).... At some point, could one of you double check to see if this is a problem on 0.15.x? I want to make sure, just since I am testing this not on an actual windows machine with IE.  But, with my virtual browser, I am getting some weirdness with buttons in general (see below -- I clicked on each button that I hover over, but none of those clicks caused anything to happen 🤪), so I'm not sure quite what is going on.

It's not urgent, since this won't be merged right away.

internet-explorer-buttons

File: kolibri/plugins/learn/assets/src/views/LibraryAndChannelBrowserMainContent.vue [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        default: 'card',
+        validator(value) {
+          return ['card', 'list'].includes(value);
+        },
+      },
+      gridType: {
+        type: Number,
+        required: true,
+        default: 1,
+      },
```
Comment (nucleogenesis) [inline kolibri/plugins/learn/assets/src/views/LibraryAndChannelBrowserMainContent.vue:95]: Can you add a comment explaining what this is and what values it expects?

File: kolibri/plugins/learn/assets/test/views/library-and-channel-browser-main-content-grid.spec.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  it('smoke test', () => {
+    expect(wrapper.exists()).toBe(true);
+  });
+  describe('When the user has a medium or large screen', () => {
+    describe('When `currentCardViewStyle` is a card', () => {
+      it('displays a `CardGrid`, and within the grid, a `HybridLearningContentCard` for each content node', () => {
+        expect(wrapper.find('[data-test="non-mobile-card-grid"]').element).toBeTruthy();
+        expect(wrapper.find('[data-test="content-card"]').element).toBeTruthy();
+      });
+      it('does not display a `CardGrid` with a `ResourceCard` for each content node in a single, full-width column', () => {
```
Comment (nucleogenesis) [inline kolibri/plugins/learn/assets/test/views/library-and-channel-browser-main-content-grid.spec.js:27]: This and some of the other tests like it doesn't test that "each content node" is rendered rather it only tests that the one in your propsData.contents does. Testing two would better cover the condition and wouldn't be too hard to do manually - you could try making each card's data-test dynamically generated `:data-test="'content-card-' + idx"` and then assert that they both show up. If you want to test more than two to be even safer, then `it.each` might help.

**Review by nucleogenesis — APPROVED** [review]
Things look good here to me. Just a note on a way to make the tests a bit more robust. Let me know if you run into issues with `it.each` if you end up taking that route.

**Comment by pcenov** [comment]
Hi @marcellamaki retested with the latest builds on Windows, Ubuntu and Android and I confirm that the search issues are now fixed and there are no other issues. Awesome!
Installed the [official 0.15 build](https://learningequality.org/r/kolibri-windows-setup-latest) and the info icon is working correctly in Internet Explorer so there must be an interim change that's caused this issue there which I guess I can report separately?

File: kolibri/plugins/learn/assets/test/views/library-and-channel-browser-main-content-grid.spec.js (hunk 2 above)
Comment (marcellamaki) [reply kolibri/plugins/learn/assets/test/views/library-and-channel-browser-main-content-grid.spec.js:27]: @nucleogenesis -- thanks for the help here. I've updated the tests now that we're ready to merge into develop again. Let me know if they are similar to what you were thinking. Thanks!

Comment (nucleogenesis) [reply kolibri/plugins/learn/assets/test/views/library-and-channel-browser-main-content-grid.spec.js:27]: Yup this is great! Thanks Marcella!

**Comment by nucleogenesis** [comment]
Does this need any further testing review @marcellamaki or did you re-request my review re: the tests only?

**Comment by marcellamaki** [comment]
Just for the tests! Thanks @nucleogenesis !
